### PR TITLE
Remove ctx argument on imported functions and allow closures.

### PR DESCRIPTION
**Pull request by lachlansneff**
This allows you to use fully-fledged closures for imported functions.

```rust
let a_str = "this is a str";

let import_object = imports! {
    "env" => {
        "foo_bar" => Func::new(|other_thing| {
            println!("{}: {}", a_str, other_thing);
        },
    },
};
```

@evgenykuzyakov Does this fix the data initialization issue you were having?

### CODE REVIEW

**Comment by CryZe** [comment]
This might accidentally also work around #188

**Comment by lachlansneff** [comment]
@CryZe Could you elaborate? I didn't really change how imported function calling worked on windows.

**Comment by CryZe** [comment]
Oh yeah, I thought ctx wouldn't be passed anymore with this, but looks like it still does. Nvm then

**Comment by CryZe** [comment]
Oh yeah, if ctx / some env internally is still used as the first argument, then this still happens.

**Comment by lachlansneff** [comment]
Oh, that's a shame. I haven't had time to look into fixing the issues you found; been working on some stuff that we're not ready to show yet.

**Comment by evgenykuzyakov** [comment]
Yes. This should fix #169

**Comment by bjfish** [comment]
Adding closures would be nice, I think whatever is done should be compatible with our current ctx usage in emscripten/C API/etc.  

It's not a big deal but this PR would be a little easier to read if the renaming of `ctx` to `env` was split into a new PR. Plus we should update the naming in the example repos/blog posts/documentation to `env` if this is changed.

**Comment by Hywan** [comment]
Why do you want to remove the context argument?

**Comment by lachlansneff** [comment]
This doesn't remove *per-say*, but really just repurposes it to hold some environmental data, whether that's the vm ctx or not. Then it adds some convenient sugar around that to make it easy to use closures as imported functions.

**Comment by syrusakbary** [comment]
Closing this PR in favor of the Import State changes from #299: https://github.com/wasmerio/wasmer/commit/7b0992e44f21828a53c8ca8c06b234d24641c09f.

This new API let the user establish the context data very easily.
See example in WASI API: https://github.com/wasmerio/wasmer/blob/2aefa731a66616781f0632d756f1f68c9df831e5/lib/wasi/src/lib.rs#L33-L53

File: lib/win-exception-handler/src/exception_handling.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ extern "C" {
     #[link_name = "callProtected"]
     pub fn __call_protected(
         trampoline: Trampoline,
-        ctx: *mut Ctx,
+        env: Option<NonNull<Env>>,
```
Comment (Hywan) [inline lib/win-exception-handler/src/exception_handling.rs]: The C code doesn't handle the `Option` correctly I guess (it expects a pointer).

**Pull request closed without merging**